File: src/main/java/org/tobi29/scapes/client/gui/GuiServerSelect.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.client.gui;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.client.states.GameStateLoadMP;
import org.tobi29.scapes.client.states.scenes.SceneMenu;
import org.tobi29.scapes.connection.ConnectionType;
import org.tobi29.scapes.engine.GameState;
import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.gui.*;
import org.tobi29.scapes.engine.opengl.texture.Texture;
import org.tobi29.scapes.engine.opengl.texture.TextureCustom;
import org.tobi29.scapes.engine.opengl.texture.TextureFilter;
import org.tobi29.scapes.engine.opengl.texture.TextureWrap;
import org.tobi29.scapes.engine.server.ConnectionInfo;
import org.tobi29.scapes.engine.server.ServerInfo;
import org.tobi29.scapes.engine.utils.BufferCreator;
import org.tobi29.scapes.engine.utils.graphics.Image;
import org.tobi29.scapes.engine.utils.io.tag.TagStructure;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.SocketChannel;
import java.util.ArrayList;
import java.util.List;

public class GuiServerSelect extends GuiMenu {
    private static final Logger LOGGER =
            LoggerFactory.getLogger(GuiServerSelect.class);
    private static final byte[] CONNECTION_HEADER = ConnectionInfo.header();
    private final List<TagStructure> servers = new ArrayList<>();
    private final List<Element> elements = new ArrayList<>();
    private final GuiComponentScrollPaneViewport scrollPane;

    public GuiServerSelect(GameState state, Gui previous, GuiStyle style) {
        super(state, "Multiplayer", previous, style);
        TagStructure scapesTag =
                state.engine().tagStructure().getStructure("Scapes");
        if (scapesTag.has("Servers")) {
            servers.addAll(scapesTag.getList("Servers"));
        }
        scrollPane = new GuiComponentScrollPaneList(pane, 16, 80, 368, 290, 70)
                .viewport();
        updateServers();
        GuiComponentTextButton add =
                new GuiComponentTextButton(pane, 112, 410, 176, 30, 18, "Add");

        add.addLeftClick(event -> {
            state.remove(this);
            state.add(new GuiAddServer(state, this, style));
        });
        back.addLeftClick(event -> disposeServers());
    }

    private static String error(Throwable e) {
        String message = e.getMessage();
        if (message != null) {
            return message;
        }
        return e.getClass().getSimpleName();
    }

    private void disposeServers() {
        for (Element element : elements) {
            try {
                element.channel.close();
            } catch (IOException e) {
                LOGGER.warn("Failed to close server info socket: {}",
                        e.toString());
            }
            scrollPane.remove(element);
        }
        elements.clear();
    }

    public void updateServers() {
        disposeServers();
        for (TagStructure tagStructure : servers) {
            Element element = new Element(scrollPane, tagStructure);
            elements.add(element);
        }
    }

    @Override
    public void update(double mouseX, double mouseY, boolean mouseInside,
            ScapesEngine engine) {
        super.update(mouseX, mouseY, mouseInside, engine);
        elements.forEach(Element::checkConnection);
    }

    public void addServer(TagStructure server) {
        servers.add(server);
        TagStructure scapesTag =
                state.engine().tagStructure().getStructure("Scapes");
        scapesTag.setList("Servers", servers);
    }

    private class Element extends GuiComponentPane {
        private final GuiComponentTextButton label;
        private final ByteBuffer outBuffer, headerBuffer =
                BufferCreator.bytes(4);
        private SocketChannel channel;
        private int readState;
        private ByteBuffer buffer;

        public Element(GuiComponent parent, TagStructure tagStructure) {
            super(parent, 0, 0, 378, 70);
            String address = tagStructure.getString("Address");
            int port = tagStructure.getInteger("Port");
            InetSocketAddress socketAddress =
                    new InetSocketAddress(address, port);
            label = new GuiComponentTextButton(this, 70, 20, 180, 30, 18,
                    "Pinging...");
            label.addLeftClick(event -> state.engine().setState(
                    new GameStateLoadMP(socketAddress, state.engine(),
                            (SceneMenu) state.scene())));
            GuiComponentTextButton delete =
                    new GuiComponentTextButton(this, 260, 20, 80, 30, 18,
                            "Delete");
            delete.addLeftClick(event -> {
                servers.remove(tagStructure);
                TagStructure scapesTag =
                        state.engine().tagStructure().getStructure("Scapes");
                scapesTag.setList("Servers", servers);
                elements.remove(this);
                scrollPane.remove(this);
            });
            try {
                channel = SocketChannel.open();
                channel.configureBlocking(false);
                if (socketAddress.isUnresolved()) {
                    throw new IOException("Could not resolve address");
                }
                channel.connect(socketAddress);
            } catch (IOException e) {
                LOGGER.info("Failed connecting to server: {}", e.toString());
                readState = -1;
                label.setText(error(e));
            }
            outBuffer = BufferCreator.bytes(CONNECTION_HEADER.length + 1);
            outBuffer.put(CONNECTION_HEADER);
            outBuffer.put(ConnectionType.GET_INFO.data());
            outBuffer.rewind();
        }

        private void checkConnection() {
            try {
                switch (readState) {
                    case 0:
                        if (channel.finishConnect()) {
                            readState++;
                        }
                        break;
                    case 1:
                        int write = channel.write(outBuffer);
                        if (!outBuffer.hasRemaining()) {
                            readState++;
                        } else if (write == -1) {
                            readState = -1;
                        }
                        break;
                    case 2: {
                        int read = channel.read(headerBuffer);
                        if (!headerBuffer.hasRemaining()) {
                            buffer = BufferCreator
                                    .bytes(4 + headerBuffer.getInt(0));
                            headerBuffer.rewind();
                            buffer.put(headerBuffer);
                            readState++;
                        } else if (read == -1) {
                            readState = -1;
                        }
                        break;
                    }
                    case 3:
                        int read = channel.read(buffer);
                        if (!buffer.hasRemaining()) {
                            ServerInfo serverInfo = new ServerInfo(buffer);
                            label.setText(serverInfo.getName());
                            Image image = serverInfo.getImage();
                            ByteBuffer imageBuffer = image.buffer();
                            ByteBuffer buffer = BufferCreator
                                    .bytes(imageBuffer.remaining());
                            buffer.put(imageBuffer);
                            buffer.rewind();
                            Texture texture = new TextureCustom(image.width(),
                                    image.height(), buffer, 0,
                                    TextureFilter.NEAREST,
                                    TextureFilter.NEAREST, TextureWrap.CLAMP,
                                    TextureWrap.CLAMP);
                            new GuiComponentIcon(this, 15, 15, 40, 40, texture);
                            readState = -1;
                        } else if (read == -1) {
                            readState = -1;
                        }
                        break;
                }
            } catch (IOException e) {
                LOGGER.info("Failed to fetch server info: {}", e.toString());
                readState = -1;
                label.setText(error(e));
            }
            if (readState == -1) {
                readState = -2;
                try {
                    channel.close();
                } catch (IOException e) {
                    LOGGER.warn("Failed to close server info socket: {}",
                            e.toString());
                }
            }
        }
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/client/gui/GuiServerSelect.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.client.gui

import org.tobi29.scapes.engine.utils.logging.KLogging
import org.tobi29.scapes.client.ScapesClient
import org.tobi29.scapes.client.states.GameStateLoadMP
import org.tobi29.scapes.client.states.GameStateLoadSocketSP
import org.tobi29.scapes.client.states.scenes.SceneMenu
import org.tobi29.scapes.connection.ConnectionInfo
import org.tobi29.scapes.connection.ConnectionType
import org.tobi29.scapes.connection.ServerInfo
import org.tobi29.scapes.engine.GameState
import org.tobi29.scapes.engine.graphics.TextureFilter
import org.tobi29.scapes.engine.graphics.TextureWrap
import org.tobi29.scapes.engine.gui.*
import org.tobi29.scapes.engine.resource.Resource
import org.tobi29.scapes.engine.server.*
import org.tobi29.scapes.engine.utils.io.ByteBuffer
import org.tobi29.scapes.engine.utils.IOException
import org.tobi29.scapes.engine.utils.tag.MutableTagList
import org.tobi29.scapes.engine.utils.tag.TagMap
import org.tobi29.scapes.engine.utils.tag.listMut
import org.tobi29.scapes.engine.utils.tag.toMap
import java.nio.channels.SelectionKey

class GuiServerSelect(state: GameState,
                      previous: Gui,
                      private val scene: SceneMenu,
                      style: GuiStyle) : GuiMenuDouble(state, "Multiplayer",
        "Add", "Back", previous, style) {
    private val servers: MutableTagList
    private val scrollPane: GuiComponentScrollPaneViewport

    init {
        val scapes = state.engine.game as ScapesClient
        servers = scapes.configMap.listMut("Servers")

        scrollPane = pane.addVert(16.0, 5.0, -1.0, -1.0) {
            GuiComponentScrollPane(it, 70)
        }.viewport

        save.on(GuiEvent.CLICK_LEFT) { event ->
            state.engine.guiStack.add("10-Menu",
                    GuiAddServer(state, this, style))
        }
        updateServers()
    }

    fun updateServers() {
        scrollPane.removeAll()
        servers.asSequence().mapNotNull { it.toMap() }.forEach { tagMap ->
            scrollPane.addVert(0.0, 0.0, -1.0, 70.0) {
                Element(it, tagMap)
            }
        }
    }

    fun addServer(server: TagMap) {
        servers.add(server)
    }

    private inner class Element(parent: GuiLayoutData,
                                tagMap: TagMap) : GuiComponentGroupSlab(
            parent) {
        val icon: GuiComponentIcon
        val label: GuiComponentTextButton
        val address: RemoteAddress

        init {
            icon = addHori(15.0, 15.0, 40.0,
                    -1.0) { parent: GuiLayoutData -> GuiComponentIcon(parent) }
            label = addHori(5.0, 20.0, -1.0, -1.0) {
                button(it, "Pinging...")
            }
            val delete = addHori(5.0, 20.0, 80.0, -1.0) {
                button(it, "Delete")
            }

            selection(label, delete)

            address = RemoteAddress(tagMap)
            label.on(GuiEvent.CLICK_LEFT) { event ->
                state.engine.switchState(GameStateLoadMP(address, state.engine,
                        scene))
            }
            delete.on(GuiEvent.CLICK_LEFT) { event ->
                servers.remove(tagMap)
                scrollPane.remove(this)
            }

            val fail = { e: Exception ->
                label.setText(error(e))
            }
            (state.engine.game as ScapesClient).connection.addConnection { worker, connection ->
                val channel = try {
                    connect(worker, address)
                } catch (e: Exception) {
                    fail(e)
                    return@addConnection
                }
                try {
                    channel.register(worker.joiner.selector,
                            SelectionKey.OP_READ)
                    label.setText("Fetching info...")
                    // Ignore invalid certificates because worst case
                    // server name and icon get faked
                    val ssl = SSLProvider.sslHandle { true }
                    val bundleChannel = PacketBundleChannel(address,
                            channel,
                            state.engine.taskExecutor, ssl, true)
                    val output = bundleChannel.outputStream
                    output.put(ConnectionInfo.header())
                    output.put(ConnectionType.GET_INFO.data())
                    bundleChannel.queueBundle()
                    if (bundleChannel.receive()) {
                        return@addConnection
                    }
                    val infoBuffer = ByteBuffer(
                            bundleChannel.inputStream.remaining())
                    bundleChannel.inputStream[infoBuffer]
                    infoBuffer.flip()
                    val serverInfo = ServerInfo(infoBuffer)
                    label.setText(serverInfo.name)
                    val image = serverInfo.image
                    val texture = state.engine.graphics.createTexture(
                            image, 0, TextureFilter.NEAREST,
                            TextureFilter.NEAREST,
                            TextureWrap.CLAMP,
                            TextureWrap.CLAMP)
                    icon.texture = Resource(texture)
                    bundleChannel.aClose()
                } catch (e: Exception) {
                    label.setText(error(e))
                } finally {
                    try {
                        channel.close()
                    } catch (e: IOException) {
                        GameStateLoadSocketSP.logger.warn(
                                e) { "Failed to close socket" }
                    }
                }
            }
        }
    }

    companion object : KLogging() {
        private fun error(e: Throwable): String {
            val message = e.message
            if (message != null) {
                return message
            }
            return e::class.java.simpleName
        }
    }
}
